Add force offset support to ApplyLinkWrench system and to Link API

Henrique-BO opened the pull request:
<!--
Please remove the appropriate section.
For example, if this is a new feature, remove all sections except for the "New feature" section

If this is your first time opening a PR, be sure to check the contribution guide:
https://gazebosim.org/docs/all/contributing
-->

# 🎉 New feature

- Builds on top of #1593

## Summary
Adds support for specifying the application point of a force in the ApplyLinkWrench system. This is done through an offset expressed in the link frame (thus, relative to the link origin).

This uses the `force_offset` field that was already present in the Wrench message, but wasn't being used in the system. Also, an overloaded method `AddWorldWrench` was added to the Link interface. 

## Test it
<!--Explain how reviewers can test this new feature manually.-->

Open a test world that has the ApplyLinkWrench system:

```
gz sim src/gz-sim/examples/worlds/apply_link_wrench.sdf
```

Send a message to the `/world/<world_name>/wrench` topic:

```
gz topic -t "/world/apply_link_wrench/wrench" -m gz.msgs.EntityWrench  -p "entity: {name: 'cylinder::link', type: LINK}, wrench: {force: {z: 100}, force_offset: {x: 0, y: 100, z: 0}}"
```

![Screencast from 30-06-2023 18_08_34](https://github.com/gazebosim/gz-sim/assets/49255495/fd11d5b8-5f68-416c-9f30-da58edf97e02)


## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

ahcorde reviewed (changes requested):
can you add a test to this new method?

Henrique-BO commented:
Added a test in 34101f7. Since there already was a `LinkAddWorldForce` test doing very similar checks and setup, I just added a few lines to the existing function. Should we keep it like this (and so maybe rename the test) or is it better to have a separate test?

azeey commented:
@osrf-jenkins run tests please

quarkytale reviewed (approved):
Works for me

azeey commented on include/gz/sim/Link.hh:
```diff
@@ -306,6 +306,19 @@ namespace gz
+      /// \param[in] _offset The point of application of the force expressed
+      /// in the link-fixed frame and relative to the center of mass.
+      /// \param[in] _torque Torque to be applied expressed in world coordinates
+      public: void AddWorldWrench(EntityComponentManager &_ecm,
```
The `AddWorldWrench` above applies the force at the link origin. In this overload, the offset is relative to the COM.  It might be a good idea to change the name to reflect that. Maybe `AddWorldWrenchRelativeToCOM`? /cc @scpeters.

Henrique-BO replied:
Renamed the function in 361cd4f

azeey commented on include/gz/sim/Link.hh:
```diff
@@ -306,6 +306,20 @@ namespace gz
+      /// is applied for one simulation step.
+      /// \param[in] _ecm Mutable Entity-component manager.
+      /// \param[in] _force Force to be applied expressed in world coordinates
+      /// \param[in] _offset The point of application of the force expressed
```
I'm not trying to be nitpicky, but the frame of expression for these quantities has always been a point of confusion, so I want to be a little more careful. I see that the offset is relative to the COM, but it's expressed in the link frame. Wouldn't it be more consistent if it was expressed in the COM frame? i.e, if you have the following SDF
```xml
<model name="test">
  <link name="base">
    <pose>1 0 0   1.57 0 0</pose>
    <inertial>
      <pose>0 1 0    0 1.57 0</pose>
      ...
      </inertial>
   </link>
</model>

With the current function, we'd be ignoring the pose of the `<inertial>` tag.

Henrique-BO replied:
I agree, expressing the offset in the COM frame is more intuitive and consistent. Changed it in 336b119.

azeey commented on src/systems/apply_link_wrench/ApplyLinkWrench.cc:
```diff
@@ -80,16 +80,17 @@ class gz::sim::systems::ApplyLinkWrenchPrivate
 /// \param[in] _msg Entity message. If it's a link, that link is returned. If
 /// it's a model, its canonical link is returned.
 /// \param[out] Force to apply.
+/// \param[out] Offset of the force application point relative to the
```
This implies that the offset in `msgs::EntityWrench` and by association, in the `component::ExternalWorldWrench`. is relative to the COM, but the point of application is the link origin. Am I right? If so, I think we should avoid having two different reference points. If the point of application is the link origin, the offset should also be from the link origin.

Henrique-BO replied:
Changed in ee693bc so that the Link API and the ApplyLinkWrench system interpret the offset in the link frame. This way, the semantics of the offset are consistent between them, the `msgs::EntityWrench` message and the `component::ExternalWorldWrench` component.

Henrique-BO commented on include/gz/sim/Link.hh:
```diff
@@ -306,6 +306,19 @@ namespace gz
+      /// \param[in] _torque Torque to be applied expressed in world coordinates
+      public: void AddWorldWrench(EntityComponentManager &_ecm,
+                                const math::Vector3d &_force,
+                                const math::Vector3d &_offset,
```
Done in f36ac69

azeey commented:
The windows node was killed after building for 3 hours (https://build.osrfoundation.org/job/ign_gazebo-pr-win/5927/). Going to merge without it. /cc @j-rivero.
